result_set.go

```diff
@@ -0,0 +1,674 @@
+		return &ResultSet{
+			resp:            resp,
+			columnNames:     colNames,
+			colNameIndexMap: colNameIndexMap,
+		}
+	}
+	for i, name := range resp.Data.ColumnNames {
+		colNames = append(colNames, string(name))
+		colNameIndexMap[string(name)] = i
+	}
+
+	return &ResultSet{
+		resp:            resp,
+		columnNames:     colNames,
+		colNameIndexMap: colNameIndexMap,
+	}
+}
+
+func genValWarps(row *nebula.Row) ([]*ValueWrapper, error) {
+	if row == nil {
+		return nil, fmt.Errorf("Failed to generate valueWrapper: invalid row")
+	}
+	var valWraps []*ValueWrapper
+	for _, val := range row.Values {
+		if val == nil {
+			return nil, fmt.Errorf("Failed to generate valueWrapper: value is nil")
+		}
+		valWraps = append(valWraps, &ValueWrapper{val})
+	}
+	return valWraps, nil
+}
+
+func genNode(vertex *nebula.Vertex) (*Node, error) {
+	if vertex == nil {
+		return nil, fmt.Errorf("Failed to generate Node: invalid vertex")
+	}
+	var tags []string
+	nameIndex := make(map[string]int)
+
+	// Iterate through all tags of the vertex
+	for i, tag := range vertex.GetTags() {
+		name := string(tag.Name)
+		// Get tags
+		tags = append(tags, name)
+		nameIndex[name] = i
+	}
+
+	return &Node{
+		vertex:          vertex,
+		tags:            tags,
+		tagNameIndexMap: nameIndex,
+	}, nil
+}
+
+func genRelationship(edge *nebula.Edge) (*Relationship, error) {
+	if edge == nil {
+		return nil, fmt.Errorf("Failed to generate Relationship: invalid edge")
+	}
+	return &Relationship{
+		edge: edge,
+	}, nil
+}
+
+func genPathWrapper(path *nebula.Path) (*PathWrapper, error) {
+	if path == nil {
+		return nil, fmt.Errorf("Failed to generate Path Wrapper: invalid path")
+	}
+	var (
+		nodeList         []*Node
+		relationshipList []*Relationship
+		segList          []segment
+		edge             *nebula.Edge
+		segStartNode     *Node
+		segEndNode       *Node
+		segType          nebula.EdgeType
+	)
+	src, err := genNode(path.Src)
+	if err != nil {
+		return nil, err
+	}
+	nodeList = append(nodeList, src)
+
+	for _, step := range path.Steps {
+		dst, err := genNode(step.Dst)
+		if err != nil {
+			return nil, err
+		}
+		nodeList = append(nodeList, dst)
+		// determine direction
+		stepType := step.Type
+		if stepType > 0 {
+			segStartNode = src
+			segEndNode = dst
+			segType = stepType
+		} else {
+			segStartNode = dst // switch with src
+			segEndNode = src
+			segType = -stepType
+		}
+		edge = &nebula.Edge{
+			Src:     []byte(segStartNode.GetID()),
+			Dst:     []byte(segEndNode.GetID()),
+			Type:    segType,
+			Name:    step.Name,
+			Ranking: step.Ranking,
+			Props:   step.Props,
+		}
+		relationship, err := genRelationship(edge)
+		if err != nil {
+			return nil, err
+		}
+		relationshipList = append(relationshipList, relationship)
+
+		segList = append(segList, segment{
+			startNode:    segStartNode,
+			relationship: relationship,
+			endNode:      segEndNode,
+		})
+		src = dst
+	}
+	// var temp []string
+	for i := 0; i < len(segList)-1; i++ {
```

Comment:
why use new `for` to check the path is right. You can check before adding a segment.